### PR TITLE
Marathon 1.5

### PR DESCRIPTION
**Pull request by katallaxie**
Adds to PR https://github.com/gambol99/go-marathon/pull/327

- Adds suggestions by @gambol99 in https://github.com/gambol99/go-marathon/pull/327
- Fixes linting errors
- Improved https://goreportcard.com/report/github.com/katallaxie/go-marathon

### CODE REVIEW

**Review by timoreimann — CHANGES_REQUESTED** [review]
Thanks for picking up the work!

Could you please move the dep-related changes into a separate PR? It requires a bit more before I'd consider ourselves to be "dep ready" (namely, extension of the Makefile and improvements on the README). While I definitely like to use dep to improve test stability, I'd rather not have the effort block progress on the Marathon 1.5 support.

**Comment by katallaxie** [comment]
👍 I have removed it. I can readly add it after this PR.

File: .gitignore [hunk 1]
```diff
@@ -20,6 +20,7 @@ coverage
 # Folders
 _obj
 _test
+vendor
```
Comment (timoreimann) [inline .gitignore]: For the sake of completion, this should still be undone.

**Review by timoreimann — CHANGES_REQUESTED** [review]

File: .gitignore (hunk 1 above)
Comment (katallaxie) [reply .gitignore]: 😃

File: README.md [hunk 2]
```diff
@@ -1,5 +1,6 @@
 [![Build Status](https://travis-ci.org/gambol99/go-marathon.svg?branch=master)](https://travis-ci.org/gambol99/go-marathon)
 [![GoDoc](http://godoc.org/github.com/gambol99/go-marathon?status.png)](http://godoc.org/github.com/gambol99/go-marathon)
+[![Go Report Card](https://goreportcard.com/badge/github.com/katallaxie/go-marathon)](https://goreportcard.com/report/github.com/katallaxie/go-marathon)
```
Comment (timoreimann) [inline README.md]: This points at the wrong repo. But really, I'd rather have this added by a separate PR as well.

Remember that if we have to revert the final commit that will end up in master for whatever functional reason, all drive-by changes will be gone as well.

**Review by timoreimann — CHANGES_REQUESTED** [review]
Some more selected notes. More importantly to me, however, are the remarks I left at https://github.com/gambol99/go-marathon/pull/327#pullrequestreview-80343301.

I am also not clear if the changes we make are 100% backwards compatible with Marathon 1.4. Ideally, we can maintain API compatibility so users coming from 1.4 don't need to change anything. But I'm not sure if that's realistic to achieve. (I haven't digged into the concrete version delta yet.)

File: README.md (hunk 2 above)
Comment (katallaxie) [reply README.md]: Lets keep it for now :)

File: application_test.go [hunk 3]
```diff
@@ -43,31 +43,70 @@ func TestApplicationMemory(t *testing.T) {
...
-		Bridged().
-		Expose(80).
-		Expose(443)
-	app, err := app.CheckHTTP("/health", 80, 5)
-	assert.Nil(t, err)
-
-	expectedAppJSONBytes, err := ioutil.ReadFile("tests/app-definitions/TestApplicationString-output.json")
-	if err != nil {
-		panic(err)
+	type test struct {
```
[the remark this answers is not in the document]
Comment (katallaxie) [reply application_test.go]: ok, it was just for sake of clarity.

File: queue.go [hunk 4]
```diff
@@ -32,7 +32,7 @@ type Item struct {
 	Application Application `json:"app"`
 }
 
-// Delay cotains the application postpone infomation
+// Delay cotains the application postpone information
```
[the remark this answers is not in the document]
Comment (katallaxie) [reply queue.go]: Fixed

File: group_test.go [hunk 5]
```diff
@@ -29,38 +29,72 @@ func TestGroups(t *testing.T) {
...
+		for _, network := range *app.Networks {
+			assert.Equal(t, "container/bridge", network.Mode)
+		}
+		if len(*app.Container.PortMappings) == 0 {
 			t.Fail()
 		}
 	}
 }
+
+// TODO @kamsz: How to work with old and new endpoints from methods.yml?
```
[the remark this answers is not in the document]
Comment (katallaxie) [reply group_test.go]: We could just leave them, because in 1.5 they are moved to a canonical representation when they are used.

[the remark this answers is not in the document]
Comment (timoreimann) [reply group_test.go]: Not perfectly sure I understand. What do you mean by _canonical representation_?

[the remark this answers is not in the document]
Comment (katallaxie) [reply group_test.go]: Ok, I just meant we could leave the fields in the fakeendpoints and test for both. Because in 1.5 they are anyway moved to a [canonical representation](https://github.com/mesosphere/marathon/blob/master/changelog.md) as they are used.

[the remark this answers is not in the document]
Comment (katallaxie) [reply group_test.go]: So, how do you suggest to proceed?

[the remark this answers is not in the document]
Comment (timoreimann) [reply group_test.go]: IIUC, with Marathon 1.5 we will continue to be able to write in the old format, while reading only returns the canonical representation (with deprecated fields being dropped).

In that case, I think we should have tests for reading both the old format (for go-marathon clients that haven't upgrade to Marathon 1.5+ yet) and the new format. At some point in the future, we can deprecate the old format.

WDYT?

**Comment by timoreimann** [comment]
Hey @katallaxie, any updates available?

Apart from the [question](https://github.com/gambol99/go-marathon/pull/335#discussion_r164897394) on how we should test functionality against pre-Marathon 1.5 and 1.5+, respectively, my other request to transform tests into table-driven forms should be more easily feasible.

Let me know your thoughts and/or questions.

**Comment by alenkacz** [comment]
is there a way I can help to move this forward?

**Comment by timoreimann** [comment]
From my point of view, when trying to finish this PR we should strive for backwards compatibility: Consumers upgrading to a Marathon 1.5-compatible version of go-marathon should ideally not need to make any code changes, regardless of whether they use Marathon 1.4 or 1.5.

If backwards compatibility is very hard to achieve in one aspect or another, we can discuss whether we can introduce a new API, break an existing API to force consumers to adjust, or come up with another resolution. All things equal, an approach maintaining backwards compatibility should be preferred.

**Comment by alenkacz** [comment]
@timoreimann After looking at the code, these are my thoughts...

- keep old and new set and get methods for networks and ports
- by using either of them, set both variants of fields (1.4 and 1.5) - make sure 1.4 can accept jsons with new fields added in 1.4
— so e.g. Application.SetNetwork would set network in both old and new way
- potentially we can read Marathon version out of info endpoint and use it to fill one or the other field based on that version (which is probably a cleaner solution)

Sounds good?

**Comment by timoreimann** [comment]
@alenkacz sounds good to me.

Reading the version and applying the right version-specific functionality would indeed enable us to minimize code complexity. I'm trying to think of ways where this may not work though... supposedly, the approach can't handle the case where a Marathon 1.4 cluster is rolling upgraded to 1.5, and consumers expect their (already upgraded) go-marathon clients to continue to work without a restart? For high availability scenarios, this could be too much of a problem, and we might need to stick to the idea of making all API calls dual-supportive. WDYT?

**Comment by alenkacz** [comment]
@timoreimann good point, but I would not be concerned about that. With minor updates between versions there are usually zookeeper migrations involved so you would not want to have both old and new versions running at the same time... Also the suggested update guide in the docs https://mesosphere.github.io/marathon/docs/upgrade/index.html confirms that

**Comment by timoreimann** [comment]
alenkacz the _Upgrading an HA installation_ in that document describes that you should run both old and new Marathon instances at one point in time. This makes a lot of sense to me as certain kinds of environments may not be able to tolerate downtime. (FWIW, that also happens to be how my organization has been doing the upgrade procedure for the past ~1.5 years.)

Coupling go-marathon's runtime behavior to the Marathon version also means that you possibly cannot switch back and forth between versions that easily. This is important when you find your new Marathon version to be broken for you, and you want to roll back to a previous version without having to restart a potential high number of go-marathon clients.

In summary, I believe it's worth trying to populate and read structs for both versions statically. If we find out that it's not possible for one reason or another, we can look for alternatives; my preference then would go towards pushing the problem back to the consumer so he can deal with it.

**Comment by alenkacz** [comment]
@timoreimann on a second look I unfortunately don't think there is a good way to do this in a backward compatible manner :( Currently this client API is very tightly coupled with the JSON so there are methods like `ExposePort(docker *Docker)` called on a Docker object, while 1.5 requires similar thing but on an object one level higher `ExposePort(container *Container)`. From Container we could set both docker and container, but from Docker we cannot set anything on container (so we cannot set new 1.5 property from current client method).

Because of that, there will never be a smooth update from 1.4 to 1.5 in terms of this client (if your client code uses the API features that were changed, you will have to change your code).

Few things come to my mind:
- keep both new and old portmappings in the application structure (basically how this PR looks right now) - which means that when you update from 1.4 to 1.5 your code will still compile but you won't be getting the data you are expecting from the methods you are using (this is the easiest way, but from a user perspective not very nice way)
- drop old methods so e.g. drop `ExposePort(docker *Docker)` and have only `ExposePort(container *Container)` and then based on marathon version set old or new way or both as we discussed (I will have to check how Marathon behaves when both old and new are set in terms of normalization - I would hope that old will get ignored) - this way when users update, their code won't compile BUT will work smoothly when they update their cluster (I like this way more)
- more options... ? I'll keep thinking :)

**Comment by timoreimann** [comment]
@alenkacz thanks for the update. I'm still somewhat worried about the approach to automatically adjust API calls based on the returned Marathon version. HTTP doesn't know sessions, so technically we won't lose a connection when hosts are swapped out behind the scene unless IP addresses are used _and_ changed, right? (In my organization, we have been pointing go-marathon at DNS records only that we map to new hosts when we upgrade.)

If automatic conversion is not feasible, then I'm inclined to say we go with what this PR provides already. It's been a while since I have looked at it,  but I'd be okay if it either populated both 1.4 and 1.5 fields from the new 1.5 API calls; or if it only populated the 1.5 fields from the 1.5 calls, requiring the user to populate the 1.4 fields instead. We can then first try to get this version out of the door and always iterate on it if we feel like it's worthwhile. I'd only like to make sure we don't break the API unnecessarily, unless there's a strong reason to. IMHO, users running on pre-1.5 shouldn't be "punished" by API changes they may not be affected by for potentially a long time.

WDYT?

**Comment by alenkacz** [comment]
@timoreimann yeah I got to the same conclusion looking at the code - since getting version before the request means doing two separate request, we are not guaranteed that it will be reliable (not talking about the network overhead).

What I don't like about the current solution is that 
- I have client against 1.4 setting network via old way and reading it from the old fields
- then i switch to 1.5
- I can still submit the apps through the old way (with 1.4 fields) but when I read from the API, the network information is accessible through a different fields and my old code suddenly does not work and I am informed about this in the runtime...

The whole problem is that the 1.5 networking changes were kind of messy and not backward compatible at all... so if we could hide this mess at least from the clients of this library, that would be nice :) I'll try something and we'll see. My 3 week vacation starts on Tuesday so we'll see how far I get :)

**Comment by timoreimann** [comment]
@alenkacz I guess you could still maintain read compatibility by making your client code check if a Marathon 1.5 field is populated, and if it is not fall back to the 1.4 field. It takes some extra logic outside of the library, but would be a good first step for users willing to pay a little price in order to use the library against Marathon 1.5 rather sooner than later.

**Comment by alenkacz** [comment]
@timoreimann yeah sure, if there is a big push to releasing that I guess what's in this branch is better than nothing - even though it is not the most pleasant client-side experience :)
